src/shims/foreign_items.rs

```diff
@@ -211,6 +214,89 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 this.machine.static_roots.push(ptr.alloc_id);
             }
 
+            // Obtains a Miri backtrace. See the README for details.
+            "miri_get_backtrace" => {
+                let tcx = this.tcx;
+                let mut data = Vec::new();
+                for frame in this.active_thread_stack().iter().rev() {
+                    data.push((frame.instance, frame.current_span().lo()));
+                }
+
+                let ptrs: Vec<_> = data.into_iter().map(|(instance, pos)| {
+                    let mut fn_ptr = this.memory.create_fn_alloc(FnVal::Instance(instance));
+                    fn_ptr.offset = Size::from_bytes(pos.0);
+                    Scalar::Ptr(fn_ptr)
+                }).collect();
+
+                let len = ptrs.len();
+
+                let ptr_ty = tcx.mk_ptr(TypeAndMut {
+                    ty: tcx.types.unit,
+                    mutbl: Mutability::Mut
+                });
+
+                let array_ty = tcx.mk_array(ptr_ty, ptrs.len().try_into().unwrap());
+                let array_ty_and_env = ParamEnv::empty().and(array_ty);
+
+                // Write pointers into array
+                let alloc = this.allocate(tcx.layout_of(array_ty_and_env).unwrap(), MiriMemoryKind::Rust.into());
+                for (i, ptr) in ptrs.into_iter().enumerate() {
+                    let place = this.mplace_index(alloc, i as u64)?;
+                    this.write_immediate_to_mplace(ptr.into(), place)?;
+                }
```

Comment:
We really need a helper for creating allocations of aggregates and immediately filling them. I think there are various copies of this code spread around miri by now

Reply:
We actually already have a helper:

https://github.com/rust-lang/miri/blob/639057e04e9286147d9896fb5cd228f0f7c19ca6/src/helpers.rs#L356

But it might not be general enough.

Reply:
Because of the reliance on field names and not field order, this new API cannot use any of the existing helpers -- and indeed I think there is little hope for it to ever share code with "normal" shims because those do rely on `repr(C)`. (Otherwise we'd have to rely on how exactly the client defines the types involved in the C call, which seems odd.)

So I wonder, maybe it would be better to make this shim behave more like a normal shim, so that it can share more of that code?

Reply:
@Aaron1011 what do you think about this?

Reply:
From the caller side, I think it's a little weird to need to return a `#[repr(C)]` struct from an `extern "Rust"` function. However, I don't think it's any weirder than my current approach, so I'm okay with changing it if it simplifies the implementation.

Reply:
After trying to implement this, it doesn't seem that we really re-use any more code by making this a `#[repr(C)]` struct. While we can use `write_packed_immediate`, we now need to construct the `Box<[u8]>` type so that we can make an `ImmTy`.

Reply:
Hm, yeah we'd want a helper that also supports scalars. But of course scalars only work if the layout is determined otherwise, which means we anyway do rely on the client to not, e.g., wrap `MiriFrame` in a newtype (which would usually be fine for normal extern calls).

Maybe let's leave introducing a common helper for the future. We should open an issue and collect which instances of this pattern we currently have in the codebase, to gather design constraints for the helper.

> From the caller side, I think it's a little weird to need to return a #[repr(C)] struct from an extern "Rust" function.

I mean, feel free to make it `extern "C"`.^^
But it doesn't actually seem that strange to me, you can link Rust code with `extern "Rust"` together if both files were compiled with the same compiler, but even then two separate copies of the same struct definition can have different layout.

Reply:
@RalfJung: If we're leaving the code as-is for now, is it fine to leave it as a non-`#[repr(C)]` type, or should I change that?

Reply:
This seems easier to remove than add in the future, so I would be in favor of adding it.

Reply:
I've added a check that `#[repr(C)]` is applied.

Reply:
Nice :)